### PR TITLE
fix: Remove sagemaker_job_name from hyperparameters in TrainingStep

### PR DESCRIPTION
**Pull request by staubhp**
*Issue #, if available:*
[2940](https://github.com/aws/sagemaker-python-sdk/issues/2940)

*Description of changes:*
Some framework classes add a `sagemaker_job_name` hyperparameter with a dynamic training job name. This is both inaccurate (the job name will be set when the step runs) and it breaks caching for those steps.

*Testing done:*
Unit, manual

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/workflow/steps.py [hunk 1]
```diff
@@ -301,6 +301,7 @@ def arguments(self) -> RequestType:
         )
         request_dict = self.estimator.sagemaker_session._get_train_request(**train_args)
         request_dict.pop("TrainingJobName")
+        request_dict["HyperParameters"].pop("sagemaker_job_name", None)
```
Comment (jerrypeng7773) [inline src/sagemaker/workflow/steps.py:304]: wonder if it is possible that this request dict does not have `HyperParameters` in it?

**Review by qidewenwhen — APPROVED** [review]

File: src/sagemaker/workflow/steps.py (hunk 1 above)
Comment (staubhp) [reply src/sagemaker/workflow/steps.py]: It will always be there, but let me add a safety check

File: tests/unit/sagemaker/workflow/test_step_collections.py [hunk 2]
```diff
@@ -457,7 +457,8 @@ def test_register_model_with_model_repack_with_estimator(
             assert len(request_dict["DependsOn"]) == 1
             assert request_dict["DependsOn"][0] == "TestStep"
             arguments = request_dict["Arguments"]
-            repacker_job_name = arguments["HyperParameters"]["sagemaker_job_name"]
+            assert BUCKET in arguments["HyperParameters"]["sagemaker_submit_directory"]
+            arguments["HyperParameters"].pop("sagemaker_submit_directory")
             assert ordered(arguments) == ordered(
```
Comment (shreyapandit) [inline tests/unit/sagemaker/workflow/test_step_collections.py:462]: This doesn't seem related to the code change made above? Why are we popping the submit_dir?

**Review by shreyapandit — CHANGES_REQUESTED** [review]
As in comments

File: tests/unit/sagemaker/workflow/test_step_collections.py (hunk 2 above)
Comment (staubhp) [reply tests/unit/sagemaker/workflow/test_step_collections.py:462]: This test was using the `sagemaker_job_name` hyperparameter to assert on the `sagemaker_submit_directory`. Because the former hyperparameter was removed. we don't know what the `sagemaker_submit_directory` will be exactly, so do a less precise assertion on it then pop it before doing equality check on the request.

**Review by jerrypeng7773 — APPROVED** [review]

File: tests/unit/sagemaker/workflow/test_step_collections.py (hunk 2 above)
Comment (shreyapandit) [reply tests/unit/sagemaker/workflow/test_step_collections.py:462]: Talked to Payton offline, we are doing necessary checks on submit dir and removing since it used job name to build the identifier, which we can't do anymore.

**Review by shreyapandit — APPROVED** [review]
